jax/_src/pallas/mosaic/lowering.py

```diff
@@ -3367,21 +3367,42 @@ def _semaphore_read_lowering_rule(
     *args,
     args_tree,
 ):
-  sem_aval, _ = tree_util.tree_unflatten(args_tree, ctx.avals_in)
+  sem_aval, sem_transforms_avals = tree_util.tree_unflatten(args_tree, ctx.avals_in)
+  primitives.check_sem_avals(
+      sem_aval,
+      sem_transforms_avals,
+      "read",
+      allowed_semaphore_types={
+          tpu_core.dma_semaphore,
```

Comment:
Why are those checks now happening in the lowering rules and not in the abstract eval rules? I'm pretty sure they can still be there. Is it because abstract eval is now in the generic pallas, but they reference dma_semaphore? Neither read nor wait (the two affected ops) should be used on the DMA semaphores, so there's no need to move those checks here.

Reply:
> Neither read nor wait (the two affected ops) should be used on the DMA semaphores, so there's no need to move those checks here.

`pallas_core.dma_semaphore,` is used in `_semaphore_read_abstract_eval`. Was that a mistake?

Reply:
I moved the check back for signal and wait and let the read one in the lowering.